test/innerJoin.js

```diff
@@ -0,0 +1,25 @@
+var R = require('..');
+var eq = require('./shared/eq');
+
+
+var a = {id: 1, name: 'a'};
+var b = {id: 2, name: 'b'};
+var c = {id: 3, name: 'c'};
+var f = R.innerJoin(function(r, id) { return r.id === id; });
+
+
+describe('innerJoin', function() {
+
+  it('only returns elements from the first list', function() {
+    eq(f([a, b, c], []), []);
+    eq(f([a, b, c], [1]), [a]);
+    eq(f([a, b, c], [1, 2]), [a, b]);
+    eq(f([a, b, c], [1, 2, 3]), [a, b, c]);
+    eq(f([a, b, c], [1, 2, 3, 4]), [a, b, c]);
+  });
+
+  it('preserves the ordering of the first list', function() {
+    eq(f([c, a, b], [1, 2, 3]), [c, a, b]);
```

Comment:
It's very much intentional. We're using an ordered collection. Order is significant, since Ramda does not consider `[1, 2]` equal to `[2, 1]`. I can think of three possible rules:

  - sort `xs'` based on the order of `xs`;
  - sort `xs'` based on the order of `ys`; or
  - apply our own sort order, perhaps the default sorting function used by `Array#sort`.

The other option is to shuffle `xs'` before returning it. This would prevent users from relying on a particular ordering. It would also break referential transparency (and possibly hurt performance), so I don't think it's a reasonable option.

Sorting `xs'` based on the order of `xs` seems best to me. Users should be able to rely on this until we change our minds and make a breaking API change (which hopefully won't happen).